### DISCUSSION TITLE
[Question] Sample to host 2 domainnames on https (433)

### DISCUSSION THREAD

**Discussion by JeepNL** [dotnet/yarp#384] [General] [answered]
I would love to try out YARP on Ubuntu 20.04 and use it for 2 live .NET 5 (preview) Blazor WASM/gRPC websites (live but prototypes). 

So, not for load balancing, but just to be able to host 2 instances of Kestrel (serving different domainnames on (internal) different ports) on one VPS with YARP in front of it serving those websites on one https (433) external port with http2 for gRPC. 

Getting an A+ grade at Qualys SSL labs (and a 100% score at https://en.internet.nl/) is really important to me, and if the sample could contain code on how to achieve this with sample code from this discussion ([Cipher Suit Preference](https://github.com/dotnet/runtime/issues/30767#issuecomment-564116466)) it would be very helpful to me and imo to others to use it. 

I know (from that same issue) how to [setup](https://github.com/dotnet/runtime/issues/30767#issuecomment-565505194) `openssl.cnf` on Ubuntu

I did this previously with NGINX, but as I said I would love to try out YARP in this configuration so I could use Microsoft software only (and because YARP is pretty, pretty fast so I'm told 😉)

**Comment by Tratcher** [comment, marked as answer]
Good questions, these are all things we still need docs for (#305). Let's take this as three different parts.

A) A+ TLS config scores on linux: We're tracking that over at https://github.com/dotnet/runtime/issues/30767 as you've noticed. Its a long conversation but I think the highlights are https://github.com/dotnet/runtime/issues/30767#issuecomment-533472765 and https://github.com/dotnet/runtime/issues/30767#issuecomment-565505194. Please give those options a try and ask any questions you need on that issue. You shouldn't need to do anything YARP specific.

B) Hosting multiple domains on the same port: The TLS feature for this is called SNI and is something you configure in Kestrel. 

Kestrel has a new [config feature](https://github.com/dotnet/aspnetcore/pull/24286) for this in 5.0.0-RC1, which isn't quite out yet but you should be able to try a [daily build](https://github.com/dotnet/installer#installers-and-binaries).

If you need to do this in any prior version then you can use the [ServerCertificateSelector](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel?view=aspnetcore-3.1#listenoptionsusehttps) callback. It lets you check the incoming domain name for each new connection and provided the matching certificate.

C) Configure YARP to route by domain name: Run through the [getting started guide](https://microsoft.github.io/reverse-proxy/articles/getting_started.html) and then tweak the config like this:
```JSON
{
 "Logging": {
   "LogLevel": {
     "Default": "Information",
     "Microsoft": "Warning",
     "Microsoft.Hosting.Lifetime": "Information"
   }
 },
 "AllowedHosts": "*",
 "ReverseProxy": {
   "Routes": [
     {
       "RouteId": "route1",
       "ClusterId": "cluster1",
       "Match": {
         "Hosts": [ "PublicDomain1" ] 
       },
     },
     {
       "RouteId": "route2",
       "ClusterId": "cluster2",
       "Match": {
         "Hosts": [ "PublicDomain2" ] 
       },
     }
   ],
   "Clusters": {
     "cluster1": {
       "Destinations": {
         "cluster1/destination1": {
           "Address": "https://internal.domain.1.name.or.ip/"
         }
       }
     },
     "cluster2": {
       "Destinations": {
         "cluster2/destination1": {
           "Address": "https://internal.domain.2.name.or.ip/"
         }
       }
     }
   }
 }
}
```

Try it out and let me know how it goes.

**Comment by JeepNL** [comment]
Thank you for your answer. If I understand correctly, with the new Kestrel config feature in 5.0.0-RC1 I can configure YARP (also a Kestrel instance) to handle the TLS connections (SSL certs) for SNI/multiple domains (https://github.com/dotnet/aspnetcore/pull/24286) and configure the routing in YARP as you described in your answer above/and by following the YARP Getting Started guide? From the comments in https://github.com/dotnet/runtime/issues/30767 I know how to configure Kestrel to get an SSL Labs A+ score. I've 2 websites (prototypes / Preview 8) running with an A+ score, but now on 2 different VPS's because I stopped using NGINX.

I'm in no hurry, I can wait until RC1 is released, which will be next month I guess, as Preview 8 just has been released.

Below ( _a top/down incoming traffic presentation_) is what I want to achieve, a basic setup using one server (Ubuntu VPS) on which I'll be running 3 Kestrel instances. 1 for YARP en 2 for the different domains.

**Internet**
    |
**YARP**: [example1.com] | [example2.com] (_both ports 80/443_)
    |
**Kestrel**: _Instance 1_ [example1.com:51443] & **Kestrel**: _Instance 2_ [example2.com:52443]

(_ports 51433, 52443 are just examples_)

**Comment by Tratcher** [comment]
Sounds right.

**Comment by samsp-msft** [comment]
Should this get moved over to discussions?

**Comment by JeepNL** [comment]
@samsp-msft Oh right, I forgot about the 'discussions' option in GitHub. I should've created a discussion instead of an issue. Chris' answer and example answered what I wanted to know to get started with YARP, looking forward doing it actually, but I've to finish some other code I'm working on at the moment first. But as soon I've something to show, or maybe have another question, I'll certainly come back here (in discussions). So if you want you can close this issue and/or move it to discussions.

**Comment by JeepNL** [comment]
Hi @Tratcher,

This weekend I got finally some time free to experiment with YARP and to look into the links you provided and tried to setup YARP with Kestrel's SSL SNI feature so I can use multiple domains names on one server listening on one port.

My config: (Dev PC) Windows 10 Pro, Visual Studio Preview (latest) and (ASP).NET 6 Preview 3. (_This YARP config is meant to run, eventually, on an Ubuntu 20.04 co-located VPS, which I'm using for my Blazor WASM sites_).

YARP is listening on `https://[::]:10443` and should proxy to `https://production.localhost:5001` or `https://beta.localhost:5002` depending on the host name. (all running on one dev PC.)

Here's my  `appsettings.Development.json` (gist.github)  [appsettings.Development.json](https://gist.github.com/JeepNL/984ff44af2b897d2b6d101fdf6256371#file-appsettings-development-json) 

I'm testing it with the 'production' url. It's a Kestrel Hosted Blazor WASM site listening on `https://production.localhost:5001/ `

In YARP's command windows it says it's `Proxying to https://production.localhost:5001/` but then there's this error: `System.Net.Http.HttpRequestException: No such host is known. (production.localhost:5001)`

YARP is unable to find the host, but the host is running.

When you look at the `appsettings.Development.json` [gist](https://gist.github.com/JeepNL/984ff44af2b897d2b6d101fdf6256371#file-appsettings-development-json), can you (_perhaps, if you've got time for it, and I'm not in a hurry_) see what I'm doing wrong? 

**Screenshot** command window YARP's error messages":

![YARP](https://user-images.githubusercontent.com/39458532/116864819-c03ef780-ac08-11eb-9300-7399eeb06a8b.jpg)

Screenshot Browser YARP

![YARP-Browser](https://user-images.githubusercontent.com/39458532/116865031-262b7f00-ac09-11eb-860f-0bebf2210df9.jpg)

The Blazor site is running at `https://production.localhost:5001`

**Screenshot** command window Kestrel Hosted Blazor WASM site (when I enter the url manually in the browser)

![production localhost](https://user-images.githubusercontent.com/39458532/116864936-f67c7700-ac08-11eb-9e6a-b0c56b7dcde0.jpg)


Screenshot Browser Blazor (when I enter the url manually in the browser)

![production localhost-Browsers](https://user-images.githubusercontent.com/39458532/116865035-2af03300-ac09-11eb-947f-cc716b69278b.jpg)

**Reply by Tratcher** [reply]
I don't see anything wrong with your setup. What you're getting is a DNS error. How have you set up your machine to resolve production.localhost? With the hosts file? It's strange that your browser can resolve it but HttpClient can't.

What happens if you do `ping production.localhost` on the command line?

**Reply by JeepNL** [reply]
OMG! You should see my face now ... ir's red 😉 I feel ashamed.

It's just a localhost DNS error, I totally forgot to edit the hosts file on my dev PC. In fact, I added the necessary (remote) DNS entries for the co-loc Ubuntu server and I remember thinking about subdomains on localhost, that it was something I should look at. But I forgot, and when my browser (Edge Chromium Dev) could access (_I don't know why, but that's besides the point_) the YARP instance of Kestrel, I didn't think about it anymore but thought I did something wrong in YARP's `appsettings.Development.json`.

**But I didn't even tried a PING (!)**. So yes, I feel ashamed, and I should be :) (_PING didn 't work_)

When I've got it all working I will create a GitHub repo so maybe it can help others who want to try a similar setup. But first  I need to test `app.UseHttpsRedirection()` with YARP, and see if gRPC is working correctly but I haven't tried anything yet, so no question here.

[Here's my new](https://gist.github.com/JeepNL/48c98c00a47bdfd1df0282e596c2284b) `appsettings.Development.json` gist, Localhost uses only 1 dev cert (_and for my Ubuntu Server I use a Let's Encrypt wildcard certificate_) so I could remove some redundant lines. And I renamed some values. Oh, yes, and I do not need subdomains for the websites, a different port is all what's needed.

Thank you for your reply and I will let you know when I've got it all working.

**Reply by JeepNL** [reply]
OMG! OMG! OMG!!! 

I've got a ASP.NET **Core 6.0 preview 5** **Kestrel Hosted Blazor WASM** website with **gRPC**/SQLite (_and Microsoft Identity, note: not IS4_) running behind **YARP** in production with a **Qualys SSL LABS A+** score on a **Ubuntu** 20.04 VPS. 

I develop all on Microsoft Windows 10 with Visual Studio Preview, I compile my code on the Ubuntu VPS without a single code change. 

IT. JUST. WORKS.
(_I've goosebumps now_)

I'll create a sample GitHub repo later this week. Then I also can show a working website which everybody can access and see the Qualys SSL LABS Score.

Now I'm ready for Preview 6 with W3C HTTP request/response logging, can't wait to play around with it. This is so much fun, working with (ASP).NET (Core) these days. It's really, really wonderful.

**Reply by pinkfloydx33** [reply]
@JeepNL any update on that repo? Would be helpful and greatly appreciated

**Reply by JeepNL** [reply]
You're right, I forgot to create a GitHub Repo, I'm sorry. Soon after I got this working I started to try out publishing a static Blazor WASM site on Cloudflare Pages, and got that working, and this front-end (on Cloudflare port 443) gets its data from a gRPC service on my VPS but that service doesn't need to be on port 80/443, so at that moment I didn't need to use YARP anymore.  I still want to use YARP in the future, but I didn't have any time to do any further development with it.

[Here's the ](https://gist.github.com/JeepNL/48c98c00a47bdfd1df0282e596c2284b) `appsettings.Development.json` gist I used. Instead of localhost subdomains (beta.localhost etc.) I think it's a better idea to create a wildcard (https) TLS certificate for a domain you own, and use that for any local testing of YARP with a https certificate. 

ie: 
```
localsite1.example.com
localsite2.example.com
```

and put those in your local HOSTS file 

```
127.0.0.1   localsite1.example.com
::1             localsite1.example.com
127.0.0.1   localsite2.example.com
::1              localsite2.example.com
```

**Reply by InteXX** [reply]
@JeepNL Do you have a suggestion on how I might use YARP to [map ports to controllers](https://stackoverflow.com/q/71481349)? The documentation is a bit daunting, especially for an ASP.NET 6 beginner such as myself (coming from WebForms).

**Reply by Tratcher** [reply]
@InteXX I think you can do that with a Host route constraint where the Host is `*:80` or `*:443` for example.

**Reply by InteXX** [reply]
@Tratcher 

Perfect. Good news. How do I make a host route constraint? Is the answer to that buried in [here](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-6.0) somewhere?

(BTW, I'm doing all my configuration in code, if that's relevant to your reply.)

**Reply by Tratcher** [reply]
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-6.0#host-matching-in-routes-with-requirehost

**Reply by InteXX** [reply]
@Tratcher 

Is a host route constraint the same as a [route constraint](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-6.0#route-constraints)?

**Reply by Tratcher** [reply]
In YARP you're setting it under Match:
https://github.com/microsoft/reverse-proxy/blob/723e04c2441fd96076e59ec385ec7fc123cfb5b7/src/ReverseProxy/Configuration/RouteMatch.cs#L24
https://microsoft.github.io/reverse-proxy/articles/config-files.html#configuration-structure

**Reply by InteXX** [reply]
@Tratcher 

You mentioned this might be possible using 'host route constraints.' I'm wondering whether that's synonymous with 'route constraints,' for which documentation is found [here](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-6.0#route-constraints). If so, that's good news—I've found documentation for what you're referring to.

The bad news is that the documentation is a bit sketchy on how I might use one/them to accomplish my goal.

Also: if I can pull this off without the complexity of YARP, I'd sure like to.

**Reply by Tratcher** [reply]
It's not the same as the path route constraints, it's the Host routing described here:
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-6.0#host-matching-in-routes-with-requirehost
`[Host("*:5000")]`

**Reply by InteXX** [reply]
@Tratcher 

OK, that gets me closer (to understanding). Thanks. I'll try it later today and let you know.

**Reply by InteXX** [reply]
@Tratcher 

I was able to resolve the port/controller mapping problem, for my case at least, by using a `MapGet` call in conjunction with the suggested (and slightly modified) [`Use` call](https://stackoverflow.com/a/49683595/722393):

```lang-cs
app = builder.Build();

app.Use(async (context, next) =>
{
  if (context.Connection.LocalPort == 80)
    context.Response.Headers["Content-Type"] = "text/html; charset=utf-8";
    await next();
});

app.MapGet("/", () => File.ReadAllText(@"Pages\Setup.htm")).RequireHost("*:80");
```

This works for my use case because I'm allowing only a single static page to be accessed on port 80, the user instructions for installing the Root CA. Once that task is done, everything else runs on 443 as normal.

So in the end I didn't need to map to a controller at all. In fact, with that approach I was wondering how I was going to break out of the navigation framework to display only that standalone setup page. This solves both problems.

But it was your `*:80` tip and the documentation links that got me here. Thanks.

**Reply by Tratcher** [reply]
Seems like you could consolidate those:
```
app.Use(async (context, next) =>
{
  if (context.Connection.LocalPort == 80)
  {
    context.Response.Headers["Content-Type"] = "text/html; charset=utf-8";
    await context.Response.SendFileAsync(@"Pages\Setup.htm");
    return;
  }
  await next();
});
```

or

```
app.MapGet("/", () => 
{
    HttpContext.Response.Headers["Content-Type"] = "text/html; charset=utf-8";
    return File.ReadAllText(@"Pages\Setup.htm")).RequireHost("*:80");
}).RequireHost("*:80");
```

**Reply by InteXX** [reply]
Perfect. I'll go with Door # 1  🙂

**Reply by InteXX** [reply]
@Tratcher 

Hm, it turns out I couldn't consolidate them.

With the first option, I get an exception when running on port 80:

```
fail: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
      System.InvalidOperationException: StatusCode cannot be set because the response has already started.
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ThrowResponseAlreadyStartedException(String value)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.set_StatusCode(Int32 value)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.Microsoft.AspNetCore.Http.Features.IHttpResponseFeature.set_StatusCode(Int32 value)
         at Microsoft.AspNetCore.Http.DefaultHttpResponse.set_StatusCode(Int32 value)
         at Microsoft.AspNetCore.StaticFiles.StaticFileContext.ApplyResponseHeaders(Int32 statusCode)
         at Microsoft.AspNetCore.StaticFiles.StaticFileContext.SendAsync()
         at Microsoft.AspNetCore.StaticFiles.StaticFileContext.ServeStaticFile(HttpContext context, RequestDelegate next)
         at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
         at Program.<>c.<<<Main>$>b__0_5>d.MoveNext() in D:\Application\Server\Program.cs:line 66
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
warn: Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[2]
      The response has already started, the error page middleware will not be executed.
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HMG7IBOT4FDN", Request id "0HMG7IBOT4FDN:00000002": An unhandled exception was thrown by the application.
      System.InvalidOperationException: StatusCode cannot be set because the response has already started.
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ThrowResponseAlreadyStartedException(String value)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.set_StatusCode(Int32 value)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.Microsoft.AspNetCore.Http.Features.IHttpResponseFeature.set_StatusCode(Int32 value)
         at Microsoft.AspNetCore.Http.DefaultHttpResponse.set_StatusCode(Int32 value)
         at Microsoft.AspNetCore.StaticFiles.StaticFileContext.ApplyResponseHeaders(Int32 statusCode)
         at Microsoft.AspNetCore.StaticFiles.StaticFileContext.SendAsync()
         at Microsoft.AspNetCore.StaticFiles.StaticFileContext.ServeStaticFile(HttpContext context, RequestDelegate next)
         at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
         at Program.<>c.<<<Main>$>b__0_5>d.MoveNext() in D:\Application\Program.cs:line 66
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
         at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
```
... using this code:

```
    57.  app = builder.Build();
    58.  
    59.  app.Use(async (context, next) =>
    60.  {
    61.    if (context.Connection.LocalPort == 80)
    62.    {
    63.      context.Response.Headers["Content-Type"] = "text/html; charset=utf-8";
    64.      await context.Response.SendFileAsync(@"Pages\Static\Setup.htm");
    65.    }
==> 66.    await next();
    67.  });
    68.  
    69.  // Configure the HTTP request pipeline.
    70.  if (app.Environment.IsDevelopment())
    71.  {
    72.    app.UseWebAssemblyDebugging();
    73.  }
    74.  else
    75.  {
    76.    app.UseExceptionHandler("/Error");
    77.    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
    78.    app.UseHsts();
    79.  }
    80.  
    81.  app.UseBlazorFrameworkFiles();
    82.  app.UseStaticFiles();
    83.  app.UseRouting();
    84.  app.MapRazorPages();
    85.  app.MapControllers();
    86.  app.MapFallbackToFile("index.html");
    87.  
    88.  app.Run();
```

... and with the second option, I get a compile-time error on `HttpContext.Response`:

> An object reference is required for the non-static field, method, or property 'HttpContext.Response'

It isn't the end of the world—I can just split them out again—but I thought I'd let you know.

**Reply by Tratcher** [reply]
You're missing a `return;` after line 64, you don't want the pipeline to keep running, you've already sent the response.

You're right about the second one, that was sloppy, you need to get to the HttpContext. I think you can pass into the lambda?

**Reply by InteXX** [reply]
Fetish: I like my methods to be one-way-in, one-way-out. So I used this variation:

```
if (context.Connection.LocalPort == 80)
{
  context.Response.Headers["Content-Type"] = "text/html; charset=utf-8";
  await context.Response.SendFileAsync(@"Pages\Static\Setup.htm");
}
else 
{
  await next();
}
```

Works great.

> pass into the lambda

Not sure I'm following you there. I tried `app.MapGet("/", (context) => { ... });` but no luck.

**Reply by InteXX** [reply]
```
    74.  app.MapGet("/", (context) =>
    75.  {
    76.    context.Response.Headers["Content-Type"] = "text/html; charset=utf-8";
==> 77.    return File.ReadAllText(@"Pages\Setup.htm");
    78.  }).RequireHost("*:80");
```
> Cannot implicitly convert type 'string' to 'System.Threading.Tasks.Task'	77
> Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type	77

**Reply by InteXX** [reply]
OK, we have to cast the `context` parameter:

```
74.  app.MapGet("/", (HttpContext context) =>
75.  {
76.    context.Response.Headers["Content-Type"] = "text/html; charset=utf-8";
77.    return File.ReadAllText(@"Pages\Setup.htm");
78.  }).RequireHost("*:80");
```
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-context?view=aspnetcore-6.0#use-httpcontext-from-minimal-apis

All is right with the world.
